pkg/operator/sync_common.go

```diff
@@ -78,31 +85,88 @@ func manageControllerResources(c serviceCAOperator, resourcePath string, modifie
 	return nil
 }
 
-// TODO manage rotation in addition to initial creation
 func manageSignerCA(client coreclientv1.SecretsGetter, eventRecorder events.Recorder) (bool, error) {
 	secret := resourceread.ReadSecretV1OrDie(v4_00_assets.MustAsset("v4.0.0/service-serving-cert-signer-controller/signing-secret.yaml"))
-	_, err := client.Secrets(secret.Namespace).Get(secret.Name, metav1.GetOptions{})
-	if !apierrors.IsNotFound(err) {
+
+	var certs []*x509.Certificate
+
+	existing, err := client.Secrets(secret.Namespace).Get(secret.Name, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		// Secret will need to be created
+	} else if err != nil {
 		return false, err
+	} else {
+		// Secret exists - attempt to parse its certs
+		certData := existing.Data[corev1.TLSCertKey]
+		if len(certData) > 0 {
+			certs, err = cert.ParseCertsPEM(certData)
+			if err != nil {
+				// TODO(marun) Invalid cert data will result in an
+				// error being returned by ParseCertsPEM. Is returning
+				// an error the best strategy, or would it be
+				// preferable to log the error and update the secret
+				// with the asset-sourced version?
+				return false, err
+			}
+		}
 	}
+
+	secretHasCert := len(certs) != 0 && certs[0] != nil
+	if secretHasCert {
+		existingCert := certs[0]
+		if !util.CertHalfwayExpired(existingCert) {
+			// Cert is up-to-date
+			return false, nil
+		}
+
+		// Rotation is required
+		klog.V(4).Infof("rotating signing CA")
+		keyData := existing.Data[corev1.TLSPrivateKeyKey]
+		if len(keyData) == 0 {
+			return false, fmt.Errorf("secret %s/%s does not provide a value for %q", secret.Namespace, secret.Name, corev1.TLSPrivateKeyKey)
+		}
+		key, err := keyutil.ParsePrivateKeyPEM(keyData)
+		if err != nil {
+			return false, err
+		}
+		signingCA, err := util.RotateSigningCA(existingCert, key.(*rsa.PrivateKey))
+		if err != nil {
+			return false, err
+		}
+		d := secret.Data
+		d[corev1.TLSCertKey], d[corev1.TLSPrivateKeyKey], d[api.IntermediateDataKey], d[api.BundleDataKey], err = signingCA.GetPEMBytes()
```

Comment:
You swapped Intermediate and Bundle keys here

Reply:
Good catch!  Done.